src/typechecker/FStar.TypeChecker.Util.fs

```diff
@@ -1011,26 +1011,96 @@ let check_trivial_precondition env c =
   
   ct, vc, Env.guard_of_guard_formula <| NonTrivial vc
 
-let maybe_coerce_bool_to_type env (e:term) (lc:lcomp) (t:term) : term * lcomp =
-    if env.is_pattern then e, lc else
+let coerce_with (env:Env.env) (e : term) (lc : lcomp) (f : lident) (ty : typ) : term * lcomp =
+    match Env.try_lookup_lid env f with
+    | Some _ ->
+        if Env.debug env (Options.Other "Coercions") then
+            BU.print1 "Coercing with %s!\n" (Ident.string_of_lid f);
+        let coertion = S.fvar (Ident.set_lid_range f e.pos) (Delta_constant_at_level 1) None in
+        let lc = bind e.pos env (Some e) lc (None, U.lcomp_of_comp <| S.mk_Total ty) in
+        let e = mk_Tm_app coertion [S.as_arg e] None e.pos in
+        e, lc
+    | None ->
+        Errors.log_issue e.pos (Errors.Warning_CoercionNotFound,
+                                (BU.format1 "Coercion %s was not found in the environment, not coercing."
+                                            (string_of_lid f)));
+        e, lc
+
+let maybe_coerce_lc env (e:term) (lc:lcomp) (t:term) : term * lcomp =
```

Comment:
Seems like `TcTerm` calls `maybe_coerce_lc` that does not do the phase 1 check. `maybe_coerce` does the check, but it seems like an internal function (that may be does not need to be in the interface?).

Reply:
Oh, thanks, my bad.

Reply:
Fixing this revealed that the test wasn't correct since we need `b2t` for bootstrapping the compiler, and `phase1` is false there. So I changed it to also coerce if `--ml_ish`. Should it also coerce if `--use_two_phase_tc` is false? Are we stil supporting that?